File: core/src/main/scala/kafka/log/remote/RemoteStorageManager.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kafka.log.remote

import java._
import java.io.IOException

import kafka.log.LogSegment
import org.apache.kafka.common.annotation.InterfaceStability
import org.apache.kafka.common.record.FileRecords.TimestampAndOffset
import org.apache.kafka.common.record.Records
import org.apache.kafka.common.{Configurable, TopicPartition}

/**
 * RemoteStorageManager is an interface that allows to plugin different remote storage implementations to copy and
 * retrieve the log segments.
 *
 * All these APIs are still experimental.
 */
@InterfaceStability.Unstable
trait RemoteStorageManager extends Configurable with AutoCloseable {

  /**
   * Earliest log offset if exists for the given topic partition in the remote storage. Return -1 if there are no
   * segments in the remote storage.
   *
   * @param tp
   * @return
   */
  @throws(classOf[IOException])
  def earliestLogOffset(tp: TopicPartition): Long

  /**
   * Copies LogSegment provided by [[RemoteLogManager]] for the given topic partition with the given leader epoch.
   * Returns the RDIs of the remote data. This method is invoked by the leader of topic partition.
   *
   * //todo LogSegment is not public, this will be changed with an interface which provides base and end offset of the
   * segment, log and offset/time indexes.
   *
   * @param topicPartition
   * @param logSegment
   * @return
   */
  @throws(classOf[IOException])
  def copyLogSegment(topicPartition: TopicPartition, logSegment: LogSegment, leaderEpoch: Int): util.List[RemoteLogIndexEntry]

  /**
   * List the remote log segment files of the given topicPartition.
   * The RemoteLogManager of a follower uses this method to find out the remote data for the given topic partition.
   *
   * @return List of remote segments, sorted by baseOffset in ascending order.
   */
  @throws(classOf[IOException])
  def listRemoteSegments(topicPartition: TopicPartition): util.List[RemoteLogSegmentInfo] = {
    listRemoteSegments(topicPartition, 0)
  }

  /**
   * List the remote log segment files of the specified topicPartition with offset >= minOffset.
   * The RLM of a follower uses this method to find out the remote data
   *
   * @param minOffset The minimum offset for a segment to be returned.
   * @return List of remote segments that contains offsets >= minOffset, sorted by baseOffset in ascending order.
   */
  @throws(classOf[IOException])
  def listRemoteSegments(topicPartition: TopicPartition, minOffset: Long): util.List[RemoteLogSegmentInfo]

  /**
   * Returns a List of RemoteLogIndexEntry for the given RemoteLogSegmentInfo. This is used by follower to store remote
   * log indexes locally.
   *
   * @param remoteLogSegment
   * @return
   */
  @throws(classOf[IOException])
  def getRemoteLogIndexEntries(remoteLogSegment: RemoteLogSegmentInfo): util.List[RemoteLogIndexEntry]

  /**
   * Deletes remote LogSegment file and indexes for the given remoteLogSegmentInfo
   *
   * @param remoteLogSegmentInfo
   * @return
   */
  @throws(classOf[IOException])
  def deleteLogSegment(remoteLogSegmentInfo: RemoteLogSegmentInfo): Boolean

  /**
   * Delete all the log segments for the given topic partition. This can be done by rename the existing locations
   * and delete them later in asynchronous manner.
   *
   * @param topicPartition
   * @return
   */
  @throws(classOf[IOException])
  def deleteTopicPartition(topicPartition: TopicPartition): Boolean

  /**
   * Remove the log segments which are older than the given cleanUpTillMs. Return the log start offset of the
   * earliest remote log segment if exists or -1 if there are no log segments in the remote storage.
   *
   * @param topicPartition
   * @param cleanUpTillMs
   * @return
   */
  @throws(classOf[IOException])
  def cleanupLogUntil(topicPartition: TopicPartition, cleanUpTillMs: Long): Long

  /**
   * Read up to maxBytes data from remote storage, starting from the 1st batch that is greater than or equals to the
   * startOffset. It will read at least one batch, if the 1st batch size is larger than maxBytes.
   *
   * @param remoteLogIndexEntry The first remoteLogIndexEntry that remoteLogIndexEntry.lastOffset >= startOffset
   * @param maxBytes            maximum bytes to fetch for the given entry
   * @param startOffset         initial offset to be read from the given rdi in remoteLogIndexEntry
   * @param minOneMessage       if true read at least one record even if the size is more than maxBytes
   * @return
   */
  @throws(classOf[IOException])
  def read(remoteLogIndexEntry: RemoteLogIndexEntry, maxBytes: Int, startOffset: Long, minOneMessage: Boolean): Records


  /**
   * Search forward for the first message that meets the following requirements:
   * - Message's timestamp is greater than or equals to the targetTimestamp.
   * - Message's offset is greater than or equals to the startingOffset.
   *
   * @param targetTimestamp The timestamp to search for.
   * @param startingOffset  The starting offset to search.
   * @return The timestamp and offset of the message found. Null if no message is found.
   */
  @throws(classOf[IOException])
  def findOffsetByTimestamp(remoteLogIndexEntry: RemoteLogIndexEntry,
                            targetTimestamp: Long,
                            startingOffset: Long): TimestampAndOffset

  /**
   * Release any system resources used by this instance.
   */
  def close(): Unit
}

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.server.log.remote.storage;

import org.apache.kafka.common.Configurable;
import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata.CustomMetadata;

import java.io.Closeable;
import java.io.InputStream;
import java.util.Map;
import java.util.Optional;

/**
 * This interface provides the lifecycle of remote log segments that includes copy, fetch, and delete from remote
 * storage.
 * <p>
 * Each upload or copy of a segment is initiated with {@link RemoteLogSegmentMetadata} containing {@link RemoteLogSegmentId}
 * which is universally unique even for the same topic partition and offsets.
 * <p>
 * {@link RemoteLogSegmentMetadata} is stored in {@link RemoteLogMetadataManager} before and after copy/delete operations on
 * {@link RemoteStorageManager} with the respective {@link RemoteLogSegmentState}. {@link RemoteLogMetadataManager} is
 * responsible for storing and fetching metadata about the remote log segments in a strongly consistent manner.
 * This allows {@link RemoteStorageManager} to have eventual consistency on metadata (although the data is stored
 * in strongly consistent semantics).
 * <p>
 * All properties prefixed with the config: "remote.log.storage.manager.impl.prefix"
 * (default value is "rsm.config.") are passed when {@link #configure(Map)} is invoked on this instance.
 *
 * Implement {@link org.apache.kafka.common.metrics.Monitorable} to enable the manager to register metrics.
 * The following tags are automatically added to all metrics registered: <code>config</code> set to
 * <code>remote.log.storage.manager.class.name</code>, and <code>class</code> set to the RemoteStorageManager class name.
 * <p>
 * Plugin implementors of {@link RemoteStorageManager} should throw {@link RetriableRemoteStorageException}
 * for transient errors that can be recovered by retrying. For non-recoverable errors,
 * {@link RemoteStorageException} should be thrown. This distinction allows RemoteLogManager to
 * handle retries gracefully and report metrics accurately.
 */
public interface RemoteStorageManager extends Configurable, Closeable {

    /**
     * Type of the index file.
     */
    enum IndexType {
        /**
         * Represents offset index.
         */
        OFFSET,

        /**
         * Represents timestamp index.
         */
        TIMESTAMP,

        /**
         * Represents producer snapshot index.
         */
        PRODUCER_SNAPSHOT,

        /**
         * Represents transaction index.
         */
        TRANSACTION,

        /**
         * Represents leader epoch index.
         */
        LEADER_EPOCH,
    }

    /**
     * Copies the given {@link LogSegmentData} provided for the given {@code remoteLogSegmentMetadata}. This includes
     * log segment and its auxiliary indexes like offset index, time index, transaction index, leader epoch index, and
     * producer snapshot index.
     * <p>
     * Invoker of this API should always send a unique id as part of {@link RemoteLogSegmentMetadata#remoteLogSegmentId()}
     * even when it retries to invoke this method for the same log segment data.
     * <p>
     * This operation is expected to be idempotent. If a copy operation is retried and there is existing content already written,
     * it should be overwritten, and do not throw {@link RemoteStorageException}
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment.
     * @param logSegmentData           data to be copied to tiered storage.
     * @return custom metadata to be added to the segment metadata after copying.
     * @throws RemoteStorageException          if there are any errors in storing the data of the segment.
     * @throws RetriableRemoteStorageException if the error is transient and the operation can be retried.
     */
    Optional<CustomMetadata> copyLogSegmentData(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                                LogSegmentData logSegmentData) throws RemoteStorageException;

    /**
     * Returns the remote log segment data file/object as InputStream for the given {@link RemoteLogSegmentMetadata}
     * starting from the given startPosition. The stream will end at the end of the remote log segment data file/object.
     *
     * <p><b>Override contract:</b> a {@link RemoteStorageManager} implementation must override either this
     * int-overload OR the long-overload ({@link #fetchLogSegment(RemoteLogSegmentMetadata, long)}). New plugins
     * should override the long-overload to support segments larger than 2GiB (KIP-1333). Failing to override
     * either will cause {@link UnsupportedOperationException} at runtime.
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment.
     * @param startPosition            start position of log segment to be read, inclusive.
     * @return input stream of the requested log segment data.
     * @throws RemoteStorageException          if there are any errors while fetching the desired segment.
     * @throws RemoteResourceNotFoundException the requested log segment is not found in the remote storage.
     * @deprecated Use {@link #fetchLogSegment(RemoteLogSegmentMetadata, long)} instead.
     *             Slated for removal in a future major release.
     */
    @Deprecated(since = "4.4")
    default InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                        int startPosition) throws RemoteStorageException {
        throw new UnsupportedOperationException(
            "RemoteStorageManager " + getClass().getName() + " has not overridden either " +
            "fetchLogSegment(RemoteLogSegmentMetadata, int) or fetchLogSegment(RemoteLogSegmentMetadata, long). " +
            "New plugins must override the long-overload (KIP-1333).");
    }

    /**
     * Returns the remote log segment data file/object as InputStream for the given {@link RemoteLogSegmentMetadata}
     * starting from the given startPosition. The stream will end at the end of the remote log segment data file/object.
     * <p>
     * This overload accepts a {@code long} start position to support segments larger than 2GB.
     * The default implementation delegates to {@link #fetchLogSegment(RemoteLogSegmentMetadata, int)} when the
     * position fits in an int, and throws {@link RemoteStorageException} otherwise. Implementations that support
     * segments larger than 2GiB <b>must</b> override this method to handle the full long range.
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment.
     * @param startPosition            start position of log segment to be read, inclusive.
     * @return input stream of the requested log segment data.
     * @throws RemoteStorageException          if there are any errors while fetching the desired segment, or if
     *                                         {@code startPosition > Integer.MAX_VALUE} and this method has not been
     *                                         overridden to support large segments.
     * @throws RemoteResourceNotFoundException the requested log segment is not found in the remote storage.
     */
    default InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                        long startPosition) throws RemoteStorageException {
        if (startPosition > Integer.MAX_VALUE) {
            throw new RemoteStorageException("Start position " + startPosition +
                " exceeds Integer.MAX_VALUE but this RemoteStorageManager implementation has not overridden " +
                "fetchLogSegment(RemoteLogSegmentMetadata, long). Override the long-overload to support " +
                "segments larger than 2GiB (KIP-1333).");
        }
        return fetchLogSegment(remoteLogSegmentMetadata, (int) startPosition);
    }

    /**
     * Returns the remote log segment data file/object as InputStream for the given {@link RemoteLogSegmentMetadata}
     * starting from the given startPosition. The stream will end at the smaller of endPosition and the end of the
     * remote log segment data file/object.
     *
     * <p><b>Override contract:</b> see {@link #fetchLogSegment(RemoteLogSegmentMetadata, int)}.
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment.
     * @param startPosition            start position of log segment to be read, inclusive.
     * @param endPosition              end position of log segment to be read, inclusive.
     * @return input stream of the requested log segment data.
     * @throws RemoteStorageException          if there are any errors while fetching the desired segment.
     * @throws RemoteResourceNotFoundException the requested log segment is not found in the remote storage.
     * @deprecated Use {@link #fetchLogSegment(RemoteLogSegmentMetadata, long, long)} instead.
     *             Slated for removal in a future major release.
     */
    @Deprecated(since = "4.4")
    default InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                        int startPosition,
                                        int endPosition) throws RemoteStorageException {
        throw new UnsupportedOperationException(
            "RemoteStorageManager " + getClass().getName() + " has not overridden either " +
            "fetchLogSegment(..., int, int) or fetchLogSegment(..., long, long). " +
            "New plugins must override the long-overload (KIP-1333).");
    }

    /**
     * Returns the remote log segment data file/object as InputStream for the given {@link RemoteLogSegmentMetadata}
     * starting from the given startPosition. The stream will end at the smaller of endPosition and the end of the
     * remote log segment data file/object.
     * <p>
     * This overload accepts {@code long} positions to support segments larger than 2GB.
     * The default implementation delegates to {@link #fetchLogSegment(RemoteLogSegmentMetadata, int, int)} when both
     * positions fit in an int, and throws {@link RemoteStorageException} otherwise. Implementations that support
     * segments larger than 2GB <b>must</b> override this method to handle the full long range.
     * <p>
     * <b>Plugin implementation contract:</b> see
     * {@link #fetchLogSegment(RemoteLogSegmentMetadata, long)} for the override contract.
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment.
     * @param startPosition            start position of log segment to be read, inclusive.
     * @param endPosition              end position of log segment to be read, inclusive.
     * @return input stream of the requested log segment data.
     * @throws RemoteStorageException          if there are any errors while fetching the desired segment, or if
     *                                         either position exceeds {@code Integer.MAX_VALUE} and this method has
     *                                         not been overridden to support large segments.
     * @throws RemoteResourceNotFoundException the requested log segment is not found in the remote storage.
     */
    default InputStream fetchLogSegment(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                        long startPosition,
                                        long endPosition) throws RemoteStorageException {
        if (startPosition > Integer.MAX_VALUE || endPosition > Integer.MAX_VALUE) {
            throw new RemoteStorageException("Positions (start=" + startPosition + ", end=" + endPosition +
                ") exceed Integer.MAX_VALUE but this RemoteStorageManager implementation has not overridden " +
                "fetchLogSegment(RemoteLogSegmentMetadata, long, long). Override the long-overload to support " +
                "segments larger than 2GiB (KIP-1333).");
        }
        return fetchLogSegment(remoteLogSegmentMetadata, (int) startPosition, (int) endPosition);
    }

    /**
     * Returns the index for the respective log segment of {@link RemoteLogSegmentMetadata}.
     * <p>
     * Note: The transaction index may not exist because of no transactional records.
     * In this case, it should throw a RemoteResourceNotFoundException, instead of returning {@code null}.
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment.
     * @param indexType                type of the index to be fetched for the segment.
     * @return input stream of the requested index.
     * @throws RemoteStorageException          if there are any errors while fetching the index.
     * @throws RemoteResourceNotFoundException the requested index is not found in the remote storage
     */
    InputStream fetchIndex(RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                           IndexType indexType) throws RemoteStorageException;

    /**
     * Deletes the resources associated with the given {@code remoteLogSegmentMetadata}. Deletion is considered as
     * successful if this call returns successfully without any errors. It will throw {@link RemoteStorageException} if
     * there are any errors in deleting the file.
     * <p>
     * This operation is expected to be idempotent. If resources are not found, it is not expected to
     * throw {@link RemoteResourceNotFoundException} as it may be already removed from a previous attempt.
     *
     * @param remoteLogSegmentMetadata metadata about the remote log segment to be deleted.
     * @throws RemoteStorageException          if there are any storage related errors occurred.
     * @throws RetriableRemoteStorageException if the error is transient and the operation can be retried.
     */
    void deleteLogSegmentData(RemoteLogSegmentMetadata remoteLogSegmentMetadata) throws RemoteStorageException;
}
